OCaml 4.14.1 release (#727)
ocaml/v3.ocaml.org: data/releases/4.14.1.md

---
kind: compiler
version: 4.14.1
date: 2022-12-20
intro: |
  This page describes OCaml version **4.14.1**, released on
  Dec 20, 2022. Go [here](/releases) for a list of all releases.

  This is a bug-fix release of [OCaml 4.14.0](/releases/4.14.0).
highlights: |
  - Bug fixes for 4.14.1
---

## Installation Instructions

The base compiler can be installed as an opam switch with the following commands:
```bash
opam update
opam switch create 4.14.1
```

> ### Configuration options

Reviewer: ```suggestion
### Configuration Options
```

The configuration of the installed opam switch can be tuned with the
following options:

> - ocaml-option-afl: set OCaml to be compiled with afl-fuzz instrumentation

Reviewer: ```suggestion
- ocaml-option-afl: sets OCaml to be compiled with `afl-fuzz` instrumentation
```

Should all these items be in monospace?
Like `ocaml-option-afl`: ?
Also, since it’s a singular thing, it will take the singular for of the verb. In this case “sets” instead of “set.”

Author: Since the option are technical terms rather than English words, I think it makes sense for them to be in monospace.

> - ocaml-option-bytecode-only: compile OCaml without the native-code compiler

Reviewer: ```suggestion
- ocaml-option-bytecode-only: compiles OCaml without the native-code compiler
```

> - ocaml-option-flambda: set OCaml to be compiled with flambda activated

Reviewer: ```suggestion
- ocaml-option-flambda: sets OCaml to be compiled with `flambda` activated
```

> - ocaml-option-musl: set OCaml to be compiled with musl-gcc

Reviewer: ```suggestion
- ocaml-option-musl: sets OCaml to be compiled with `musl-gcc`
```

> - ocaml-option-no-flat-float-array: set OCaml to be compiled with --disable-flat-float-array

Reviewer: ```suggestion
- ocaml-option-no-flat-float-array: sets OCaml to be compiled with `--disable-flat-float-array`
```

> - ocaml-option-static :set OCaml to be compiled with musl-gcc -static

Reviewer: ```suggestion
- ocaml-option-static :sets OCaml to be compiled with `musl-gcc -static`
```

> - ocaml-option-address-sanitizer: set OCaml to be compiled with address sanitizer

Reviewer: ```suggestion
- ocaml-option-address-sanitizer: sets OCaml to be compiled with address sanitiser
```

> - ocaml-option-leak-sanitizer: set OCaml to be compiled with leak sanitizer

Reviewer: ```suggestion
- ocaml-option-leak-sanitizer: sets OCaml to be compiled with leak sanitiser
```

For instance, one can install a switch with both `flambda` and the naked-pointer checker enabled with

```
opam switch create 4.14.1+flambda+nnpchecker --package=ocaml-variants.4.14.1+options,ocaml-option-flambda,ocaml-option-nnpchecker
```

or with opam 2.1:

```
opam switch create 4.14.1+flambda+nnpchecker ocaml-variants.4.14.1+options ocaml-option-flambda ocaml-option-nnpchecker
```

---

Source Distribution
-------------------

- [Source
  tarball](https://github.com/ocaml/ocaml/archive/4.14.1.tar.gz)

>   (.tar.gz) for compilation under Unix (including Linux and MacOS X)

Reviewer: ```suggestion
  (.tar.gz) for compilation under Unix (including Linux and macOS X)
```

  and Microsoft Windows (including Cygwin).
- Also available in
  [.zip](https://github.com/ocaml/ocaml/archive/4.14.1.zip)
  format.

> - [OPAM](https://opam.ocaml.org/) is a source-based distribution of

Reviewer: ```suggestion
- [opam](https://opam.ocaml.org/) is a source-based distribution of
```

  OCaml and many companion libraries and tools. Compilation and
  installation are automated by powerful package managers.
- The official development repo is hosted on
  [GitHub](https://github.com/ocaml/ocaml).

The
[INSTALL](https://v2.ocaml.org/releases/4.14/notes/INSTALL.adoc) file
of the distribution provides detailed compilation and installation

> instructions — see also the [Windows release

Reviewer: ```suggestion
instructions. See also the [Windows release
```

notes](https://v2.ocaml.org/releases/4.14/notes/README.win32.adoc) for
instructions on how to build under Windows.

---

## Changes in OCaml 4.14.1 (20 December 2022)

This is the
[changelog](https://v2.ocaml.org/releases/4.14/notes/Changes).

### Compiler User-Interface and Warnings:

> - [#11184](https://github.com/ocaml/ocaml/issues/11184), [#11670](https://github.com/ocaml/ocaml/issues/11670): Stop calling ranlib on created / installed libraries

Reviewer: ```suggestion
- [#11184](https://github.com/ocaml/ocaml/issues/11184), [#11670](https://github.com/ocaml/ocaml/issues/11670): Stop calling `ranlib` on created / installed libraries
```

  (Sébastien Hinderer and Xavier Leroy, review by the same)

### Build System:

- [#11370](https://github.com/ocaml/ocaml/issues/11370), [#11373](https://github.com/ocaml/ocaml/issues/11373): Don't pass CFLAGS to flexlink during configure.
  (David Allsopp, report by William Hu, review by Xavier Leroy and
   Sébastien Hinderer)

> - [#11487](https://github.com/ocaml/ocaml/issues/11487): Thwart FMA test optimization during configure

Reviewer: ```suggestion
- [#11487](https://github.com/ocaml/ocaml/issues/11487): Thwart FMA test optimisation during configure
```

  (William Hu, review by David Allsopp and Sébastien Hinderer)

### Bug Fixes:

- [#10768](https://github.com/ocaml/ocaml/issues/10768), [#11340](https://github.com/ocaml/ocaml/issues/11340): Fix typechecking regression when combining first class
  modules and GADTs.
  (Jacques Garrigue, report by François Thiré, review by Matthew Ryan)

- [#11204](https://github.com/ocaml/ocaml/issues/11204): Fix regression introduced in 4.14.0 that would trigger Warning 17 when
  calling virtual methods introduced by constraining the self type from within
  the class definition.
  (Nicolás Ojeda Bär, review by Leo White)

> - [#11263](https://github.com/ocaml/ocaml/issues/11263), [#11267](https://github.com/ocaml/ocaml/issues/11267): caml/{memory,misc}.h: check whether `_MSC_VER` is defined

Reviewer: ```suggestion
- [#11263](https://github.com/ocaml/ocaml/issues/11263), [#11267](https://github.com/ocaml/ocaml/issues/11267): `caml/{memory,misc}.h`: Check whether `_MSC_VER` is defined
```

  before using it to ensure that the headers can always be used in code which

>   turns on -Wundef (or equivalent).

Reviewer: ```suggestion
  turns on `-Wundef` (or equivalent).
```

  (David Allsopp and Nicolás Ojeda Bär, review by Nicolás Ojeda Bär and
   Sébastien Hinderer)

> - [#11314](https://github.com/ocaml/ocaml/issues/11314), [#11416](https://github.com/ocaml/ocaml/issues/11416): fix non-informative error message for module inclusion

Reviewer: ```suggestion
- [#11314](https://github.com/ocaml/ocaml/issues/11314), [#11416](https://github.com/ocaml/ocaml/issues/11416): Fix non-informative error message for module inclusion
```
Match capitalisation

  (Florian Angeletti, report by Thierry Martinez, review by Gabriel Scherer)

> - [#11358](https://github.com/ocaml/ocaml/issues/11358), [#11379](https://github.com/ocaml/ocaml/issues/11379): Refactor the initialization of bytecode threading,

Reviewer: ```suggestion
- [#11358](https://github.com/ocaml/ocaml/issues/11358), [#11379](https://github.com/ocaml/ocaml/issues/11379): Refactor the initialisation of bytecode threading,
```

  This avoids a "dangling pointer" warning of GCC 12.1.
  (Xavier Leroy, report by Armaël Guéneau, review by Gabriel Scherer)

> - [#11387](https://github.com/ocaml/ocaml/issues/11387), module type with constraints no longer crash the compiler in presence

Reviewer: ```suggestion
- [#11387](https://github.com/ocaml/ocaml/issues/11387): Module type with constraints no longer crash the compiler in presence
```
Match formatting/capitalisation of most other list items

  of both shadowing warnings and the `-bin-annot` compiler flag.
  (Florian Angeletti, report by Christophe Raffalli, review by Gabriel Scherer)

> - [#11392](https://github.com/ocaml/ocaml/issues/11392), [#11392](https://github.com/ocaml/ocaml/issues/11392): assertion failure with -rectypes and external definitions

Reviewer: ```suggestion
- [#11392](https://github.com/ocaml/ocaml/issues/11392), [#11392](https://github.com/ocaml/ocaml/issues/11392): Assertion failure with `-rectypes` and external definitions
```
Match capitalisation

  (Gabriel Scherer, review by Florian Angeletti, report by Dmitrii Kosarev)

- [#11417](https://github.com/ocaml/ocaml/issues/11417): Fix regression allowing virtual methods in non-virtual classes.
  (Leo White, review by Florian Angeletti)

- [#11468](https://github.com/ocaml/ocaml/issues/11468): Fix regression from [#10186](https://github.com/ocaml/ocaml/issues/10186) (OCaml 4.13) detecting IPv6 on Windows for

>   mingw-w64 i686 port.

Reviewer: ```suggestion
  Mingw-w64 i686 port.
```

  (David Allsopp, review by Xavier Leroy and Sébastien Hinderer)

- [#11489](https://github.com/ocaml/ocaml/issues/11489), [#11496](https://github.com/ocaml/ocaml/issues/11496): More prudent deallocation of alternate signal stack
  (Xavier Leroy, report by @rajdakin, review by Florian Angeletti)

- [#11516](https://github.com/ocaml/ocaml/issues/11516), [#11524](https://github.com/ocaml/ocaml/issues/11524): Fix the `deprecated_mutable` attribute.
  (Chris Casinghino, review by Nicolás Ojeda Bär and Florian Angeletti)

- [#11194](https://github.com/ocaml/ocaml/issues/11194), [#11609](https://github.com/ocaml/ocaml/issues/11609): Fix inconsistent type variable names in "unbound type var"
  messages
  (Ulysse Gérard and Florian Angeletti, review Florian Angeletti and
   Gabriel Scherer)

- [#11622](https://github.com/ocaml/ocaml/issues/11622): Prevent stack overflow when printing a constructor or record
  mismatch error involving recursive types.
  (Florian Angeletti, review by Gabriel Scherer)

- [#11732](https://github.com/ocaml/ocaml/issues/11732): Ensure that types from packed modules are always generalised
  (Stephen Dolan and Leo White, review by Jacques Garrigue)

> - [#11737](https://github.com/ocaml/ocaml/issues/11737): Fix segfault condition in Unix.stat under Windows in the presence of

Reviewer: ```suggestion
- [#11737](https://github.com/ocaml/ocaml/issues/11737): Fix segfault condition in `Unix.stat` under Windows in the presence of
```

  multiple threads.
  (Marc Lasson, Nicolás Ojeda Bär, review by Gabriel Scherer and David Allsopp)

- [#11776](https://github.com/ocaml/ocaml/issues/11776): Extend environment with functor parameters in `strengthen_lazy`.
  (Chris Casinghino and Luke Maurer, review by Gabriel Scherer)

> - [#11533](https://github.com/ocaml/ocaml/issues/11533), [#11534](https://github.com/ocaml/ocaml/issues/11534): follow synonyms again in #show_module_type

Reviewer: ```suggestion
- [#11533](https://github.com/ocaml/ocaml/issues/11533), [#11534](https://github.com/ocaml/ocaml/issues/11534): Follow synonyms again in #show_module_type
```
To match capitalisation in most other list items

  (this had stopped working in 4.14.0)
  (Gabriel Scherer, review by Jacques Garrigue, report by Yaron Minsky)

- [#11768](https://github.com/ocaml/ocaml/issues/11768), [#11788](https://github.com/ocaml/ocaml/issues/11788): Fix crash at start-up of bytecode programs in

>   no-naked-pointers mode caused by wrong initialization of caml_global_data

Reviewer: ```suggestion
  no-naked-pointers mode caused by wrong initialisation of `caml_global_data`
```

  (Xavier Leroy, report by Etienne Millon, review by Gabriel Scherer)

> - [#11803](https://github.com/ocaml/ocaml/issues/11803), [#11808](https://github.com/ocaml/ocaml/issues/11808): on x86, the destination of an integer comparison must be

Reviewer: ```suggestion
- [#11803](https://github.com/ocaml/ocaml/issues/11803), [#11808](https://github.com/ocaml/ocaml/issues/11808): On x86, the destination of an integer comparison must be
```
Match capitalisation

>   a register, it cannot be a stack slot.

Reviewer: ```suggestion
  a register; it cannot be a stack slot.
```
